UNIX sockets support for RPC

laanwj opened the pull request:
Add functionality for RPC over UNIX sockets, on platforms that support UNIX sockets. This is specified with `:unix:/path/to/socket` to `-rpcbind` and `-rpcconnect`. By default the socket path is relative to the data directory.

For the server side this works as-is.

For the client side (bitcoin-cli) this requires a small patch to libevent to be able to pass in a file descriptor of an existing connection (https://github.com/libevent/libevent/pull/479).

Even without the client change this could be useful though. For example the Python RPC tests could connect through UNIX socket to avoid port collisions ~~(TODO: make a Python example)~~. This has been implemented.

Also the overhead of using UNIX sockets should be lower than using local TCP (see https://stackoverflow.com/questions/14973942/performance-tcp-loopback-connection-vs-unix-domain-socket).

### Example

Server:

```bash
src/bitcoind -printtoconsole -datadir=/store/tmp/testbtc -connect=0 -debug=rpc -debug=http -rpcbind=:unix:socket
...
2017-03-05 10:29:31 Binding RPC on UNIX socket /store/tmp/testbtc/socket
...
```

Client:

```bash
src/bitcoin-cli -rpcconnect=":unix:socket" -datadir=/store/tmp/testbtc getinfo
{
  "version": 149900,
  ...
}
```

laanwj committed:
rpc: UNIX sockets support

Add functionality for RPC over UNIX sockets, on platforms that support
UNIX sockets. This is specified with `:unix:/path/to/socket` to
`-rpcbind` and `-rpcconnect`. By default the socket path is relative to the
data directory.

laanwj commented on src/Makefile.am:
```diff
@@ -377,6 +381,9 @@ bitcoind_LDADD += $(BOOST_LIBS) $(BDB_LIBS) $(SSL_LIBS) $(CRYPTO_LIBS) $(MINIUPN
 
 # bitcoin-cli binary #
 bitcoin_cli_SOURCES = bitcoin-cli.cpp
+if BUILD_EVUNIX
```
@theuni This is pretty ugly, but I don't quite know how to do this otherwise. My first intuition was to add the file to `libbitcoin_util_a_SOURCES`, however these can't depend on libevent as that is shared between all executables. There is no library that is shared between just `bitcoind` and `bitcoin-cli` and adding one just for this seems overkill.

theuni replied:
I think this is OK. I think we'll end up with a few wrappers for libevent, so moving those to their own internal helper lib will  make sense.

gmaxwell commented:
Neat.

paveljanik commented:
This does not compile with `libevent2-2.0.22` which is OK with the current master.

```
bitcoin-cli.cpp:214:13: error: use of undeclared identifier `evhttp_connection_base_bufferevent_new'
            evhttp_connection_base_bufferevent_new(base.get(), __null, bev, "", 0)
            ^
1 error generated.
```

laanwj commented:
@paveljanik Yes, that needs a version `#ifdef`, that call doesn't exist in older versions of libevent, and on newer ones (until my patch lands) it ignores the current connection and tries to open a new one so that trick doesn't work. I'd recommend testing with my patched libevent. Alternatively comment out the bitcoin-cli change and test connecting with something else:
- `nc -U <socket>` works, for example
- curl also has a ` --unix-socket` parameter, but haven't tried it yet

TheBlueMatt commented:
For merge we should probably just not have the client version, then?

Still, this is cool - lots of better auth controls can be applied to
sockets than IPs :)

On 03/04/17 21:05, Wladimir J. van der Laan wrote:
> @paveljanik <https://github.com/paveljanik> Yes, that needs a version
> |#ifdef|, that call doesn't exist in older versions of libevent, and on
> newer ones (until my patch lands) it ignores the current connection and
> tries to open a new one so that trick doesn't work. I'd recommend
> testing with my patched libevent. Alternatively comment out the
> bitcoin-cli change and test connecting with something else:
> 
>   * |nc -U <socket>| works, for example
>   * curl also has a |--unix-socket| parameter, but haven't tried it yet
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/9919#issuecomment-284180780>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnoHrbUK-rDkbYbVKJSmWq4tqSJyg64ks5ridILgaJpZM4MTHvN>.
>

laanwj commented:
> For merge we should probably just not have the client version, then?

Would prefer putting it behind `#ifdef EXPERIMENTAL_LIBEVENT` or such, with a comment that it can be replaced with the appropriate version comparison when it lands in a version. That way, people that compile their own stuff can use it already.

laanwj commented:
> Still, this is cool - lots of better auth controls can be applied to
sockets than IPs :)

Indeed. Default permissions of the socket will be `700` due to our umask setting, which seems okay:
```
srwx------  1 user user          0 Mar  5 11:40 socket
```

jonasschnelli commented:
This is great!
Concept ACK.

laanwj committed:
tests: Run RPC tests optionally over UNIX socket

laanwj commented:
Added a commit that makes the tests' RPC run over UNIX sockets on platforms that support it, except for a few tests that make explicit assumptions about running over TCP/IP.

jnewbery commented:
I've been trying to test this locally by running the qa tests. Something that I do quite often is to run the tests with --noshutdown, wait for the test to fail, and use bitcoin-cli to interactively run RPCs on the test nodes:

`bitcoin.conf -conf=<test temp directory>/node<x>/bitcoin.conf`

With this PR, I'm not able to do that anymore because the conf file doesn't contain the username/password, and even when I add them back in, I get the error:

```
/tmp/user/1000/test_6o70u2b/28404/node0
→ ~/bitcoin/src/bitcoin-cli -rpcconnect=":unix:rpc_socket" -datadir=/tmp/user/1000/test_6o70u2b/28404/node0 help
[warn] evhttp_connection_connect_: connection to "" failed: No such file or directory
error: couldn't connect to server: unknown (code -1)
(make sure server is running and you are connecting to the correct RPC port)
/tmp/user/1000/test_6o70u2b/28404/node0
→ ll
total 16
drwxrwxr-x 3 ubuntu ubuntu 4096 Mar  6 23:31 ./
drwxrwxr-x 4 ubuntu ubuntu 4096 Mar  6 23:11 ../
-rw-rw-r-- 1 ubuntu ubuntu   79 Mar  6 23:20 bitcoin.conf
drwx------ 5 ubuntu ubuntu 4096 Mar  6 23:26 regtest/
srwx------ 1 ubuntu ubuntu    0 Mar  6 23:11 rpc_socket=
```

Do you have a better way of connecting to a test node during qa testing?

laanwj commented:
> With this PR, I'm not able to do that anymore because the conf file doesn't contain the username/password, and even when I add them back in, I get the error:

Right, I could change it to add the options to the bitcoin.conf. 

> And even when I add them back in, I get the error:

The above should work, though. That's exactly how I've been testing it. The connection failure is strange. Did you patch libevent so that `bitcoin-cli` is able to connect to UNIX sockets?

laanwj committed:
tests: Add environment option to force RPC over TCP for tests

laanwj commented:
@jnewbery Okay, added two commits:
- add `BITCOIN_TEST_RPC_TCP` environment variable to force tests to use TCP instead of UNIX sockets for RPC
- write connection info to `bitcoin.conf` instead of providing it on the command line

laanwj commented:
Needed rebase, so rebased and squashed the squasme: commits.

jonasschnelli commented:
F.I.Y:
I wanted to test this via binaries built in gitian.
I could start the daemon (wanted to test in conjunction with Q): ` ./bitcoin-0.14.99/bin/bitcoin-qt -debug=rpc -debug=http -rpcbind=nix:socket --regtest`

However the CLI responded with:
`error: RPC was asked to connect to a UNIX socket. However, the version of libevent that this utility was compiled with has no support for that.`

Maybe we should bump libevent in the dependencies directory?

laanwj commented:
> However the CLI responded with:

Yes, that's expected. `bitcoin-cli` support needs this patch: https://github.com/laanwj/libevent/commit/403b793771341460b262edb1abc31fae267174bd.patch  as well as defining `-DLIBEVENT_EXPERIMENTAL`

You should still be able to use the socket with other clients, for example from Python or with curl.

laanwj commented:
Rebased (RPC tests logging changes #9768)

laanwj committed:
rpc: More robust localhost-UNIX detection

evhttp has no direct support for UNIX sockets thus
is not able to recognize UNIX peers. Add custom code in
HTTPRequest::GetPeer and evunix to recognize these connections
as coming from localhost.

The previous solution did not work on some libevent versions.

JeremyRubin commented:
Oops, I accidentally did my review of this code on the other PR. See here https://github.com/bitcoin/bitcoin/pull/9979#pullrequestreview-31071186

ryanofsky reviewed (commented):
utACK 695e854a0444bb5c81b1422494451f17db0b5f1e, though code is pretty out of date

jnewbery commented:
I've had a quick read through the changes to the functional tests, but I'll defer doing a full review until this is rebased. A few general comments:

- I'd prefer a command line argument over and environment variable for `BITCOIN_TEST_RPC_TCP`, partly because adding a command line argument means the help text will be updated and the option will be better documented.
- there's a response to your libevent PR here: https://github.com/libevent/libevent/pull/479. It'd be good to the functionality merged into libevent so we can avoid the `LIBEVENT_EXPERIMENTAL` stuff
- I also think it would be better to make the changes to the functional test framework after TestNode changes in #10082 are in. I really dislike having state for nodes floating around in util.py (eg the new `conninfo_for()` function)

theuni commented on src/httpserver.cpp:
```diff
@@ -619,6 +667,12 @@ CService HTTPRequest::GetPeer()
+            /* Special: will get here for UNIX sockets. As we have no way to indicate that,
+             * just pass localhost IPv6.
+             */
+            address = "::1";
```
+1. This seems like another good use-case for NET_INTERNAL from #10446 (or something similar). As a bonus, that would allow us to use associative containers for unix socket addresses, since each path would have its own hash.

theuni commented on src/support/evunix.cpp:
```diff
@@ -0,0 +1,107 @@
+        close(fd);
+        return -1;
+    }
+    evutil_make_socket_nonblocking(fd);
```
need to disable SIGPIPE here too?

theuni commented on src/httpserver.cpp:
```diff
@@ -505,6 +547,12 @@ void StopHTTPServer()
+#ifdef HAVE_SOCKADDR_UN
+    //! Clean up UNIX sockets on shutdown
+    for (const auto& path: cleanupUNIXSockets) {
+         evunix_remove_socket(path);
```
should probably .clear() in case this gets called twice.

theuni reviewed (commented):
Concept ACK. I added a few comments, and @ryanofsky beat me to several others.

laanwj commented:
Closing this for now. I'll probably pick it up again for 0.16.

Sjors commented:
Concept ACK. c-lightning does this too FWIW.

This was [briefly discussed](http://www.erisian.com.au/bitcoin-core-dev/log-2018-10-25.html#l-901) during the meeting today. Tag as Up for Grabs?

MarcoFalke commented:
Marked as "up for grabs"

laanwj commented:
The server side should be easy to pick up.
As for the client side I have opened a libevent issue to track this here: https://github.com/libevent/libevent/issues/907
Unfortunately there is very little interest from their side for this.

vinipsmaker commented:
Does it have to be HTTP? Would it be easier to implement if it was pure (line delimited) JSON-RPC over UNIX domain sockets?

githubnew-pro commented:
> قابلیت های RPC را از طریق سوکت های UNIX ، در سیستم عامل های پشتیبانی از سوکت های UNIX ، اضافه کنید. این با `:unix:/path/to/socket`to `-rpcbind`و مشخص شده است `-rpcconnect`. به طور پیش فرض مسیر سوکت نسبت به فهرست داده است.
> 
> برای سمت سرور این همانطور که هست کار می کند.
> 
> برای طرف مشتری (bitcoin-cli) این نیاز به یک وصله کوچک دارد تا بتواند در توصیف پرونده اتصال موجود ( [libevent / libevent # 479](https://github.com/libevent/libevent/pull/479) ) عبور کند.
> 
> حتی بدون تغییر مشتری ، این می تواند مفید باشد. به عنوان مثال آزمایش های RPC پایتون می توانند از طریق سوکت UNIX متصل شوند تا از برخورد بندر جلوگیری شود~(TODO: یک مثال پایتون بسازید)~. این اجرا شده است.
> 
> همچنین سربار استفاده از سوکتهای UNIX باید کمتر از استفاده از TCP محلی باشد (به https://stackoverflow.com/questions/14973942/performance-tcp-loopback-connection-vs-unix-domain-socket مراجعه [کنید](https://stackoverflow.com/questions/14973942/performance-tcp-loopback-connection-vs-unix-domain-socket) ).
> 
> ### مثال
> سرور:
> 
> ```shell
> src / bitcoind -printtoconsole -datadir = / store / tmp / testbtc -connect = 0 -debug = rpc -debug = http -rpcbind =: unix: سوکت
> ...
> 05/07/2017 10:29:31 اتصال RPC در سوکت / فروشگاه / tmp / testbtc / سوکت UNIX
> ...
> ```
> 
> مشتری:
> 
> ```shell
> src / bitcoin-cli -rpcconnect = " : unix: socket " -datadir = / store / tmp / testbtc getinfo
> {
>   " نسخه " : 149900 ،
>   ...
> }
> ```

michaelfolkson commented:
> Does it have to be HTTP? Would it be easier to implement if it was pure (line delimited) JSON-RPC over UNIX domain sockets?

This was discussed today on [IRC](https://gnusha.org/bitcoin-core-dev/2021-09-22.log). In summary, this is what c-lightning does for its UNIX socket support and what Core could potentially do in future for UNIX socket support as an alternative approach to getting a PR merged in libevent. Any non HTTP based UNIX socket support would need to be in addition to the existing HTTP based JSON-RPC rather than as a replacement.

ryanofsky commented:
> i don't see any problem personally with adding support for unix sockets rpc server / client that uses plain json-rpc lines, if someone wants to work on that

This seems ok, but without URLs, you would have to decide how to handle wallet endpoints. Maybe create different sockets for different wallets like `<datadir>/wallets/<walletname>/rpc.sock`, or use one socket and accept an extra wallet RPC param.

JeremyRubin commented:
> Any non HTTP based UNIX socket support would need to be in addition to the existing HTTP based JSON-RPC rather than as a replacement.

contestable, i think it would be great to deprecate then remove http support & leave it up to external proxy services.
